neps/nep-0591.md

```diff
@@ -0,0 +1,207 @@
+---
+NEP: 591
+Title: Global Contracts
+Authors: Bowen Wang <bowen@nearone.org>, Anton Puhach <anton@nearone.org>, Stefan Neamtu <stefan.neamtu@nearone.org>
+Status: Draft
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/591
+Type: Protocol
+Replaces: 491
+Version: 1.0.0
+Created: 2025-02-11
+LastUpdated: 2025-02-11
+---
+
+## Summary
+
+This proposal introduces global contracts, a new mechanism that allows smart contracts to be deployed once and reused by any account without incurring high storage costs.
+
+Currently, deploying the same contract multiple times on different accounts leads to significant storage fees.
+Global contracts solve this by making contract code available globally, allowing multiple accounts to reference it instead of storing their own copies.
+
+Rather than requiring full storage costs for each deployment, accounts can simply link to an existing global contract, reducing redundancy and improving scalability. This approach optimizes storage, lowers costs, and ensures efficient contract distribution across the network.
+
+## Motivation
+
+A common use case on NEAR is to deploy the same smart contract many times on many different accounts. For example, a multisig contract is a frequently deployed contract.
+However, today each time such a contract is deployed, a user has to pay for its storage and the cost is quite high. For a 300kb contract the cost is 3N.
+
+With the advent of chain signatures, the smart contract wallet use case will become more ubiquitous.
+As a result, it is very desirable to be able to reuse already deployed contract without having to pay for the storage cost again.
+
+Additionally global contracts cover the underlying use case for [NEP-491](https://github.com/near/NEPs/pull/491): https://github.com/near/nearcore/pull/12818.
```

Comment:
Can you add a short context and explanation?

Reply:
@staffik can you please help me here

Reply:
```suggestion
Additionally, global contracts cover the underlying use case for [NEP-491](https://github.com/near/NEPs/pull/491): [#12818](https://github.com/near/nearcore/pull/12818).
Some businesses onboard their users by creating an account on their behalf and deploying a contract to it. Such a use case is susceptible to refund abuse, where there is a financial incentive to repeatedly create and destroy an account, cashing in on the storage fee that was initially paid by the business to deploy the contract.
```
cc @pugachAG